### PR TITLE
Add gRPC client and client factory documentation

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/grpc/grpc-dotnet/issues/435

Internal review URLs:
* https://review.docs.microsoft.com/en-us/aspnet/core/grpc/client?view=aspnetcore-3.0&branch=pr-en-us-13684
* https://review.docs.microsoft.com/en-us/aspnet/core/grpc/clientfactory?view=aspnetcore-3.0&branch=pr-en-us-13684
* https://review.docs.microsoft.com/en-us/aspnet/core/grpc/configuration?view=aspnetcore-3.0&branch=pr-en-us-13684

~Note, some of the content is for preview 9. Should not be merged until there is a preview 9 branch.~

Rebased on release/3.0-preview9.

### CODE REVIEW

**Comment by JamesNK** [comment]
@shirhatti @JunTaoLuo Cleaned up, ready for review 🙏

File: aspnetcore/grpc/clientfactory.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+public class AggregatorService : Aggregator.AggregatorBase
+{
+    private readonly Greeter.GreeterClient _client;
+
+    public AggregatorService(Greeter.GreeterClient client)
+    {
+        _client = client;
+    }
+
+    public override async Task SayHellos(HelloRequest request,
```
Comment (scottaddie) [inline aspnetcore/grpc/clientfactory.md]: Should this method be called `SayHello` instead?

File: aspnetcore/grpc/clientfactory.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        IServerStreamWriter<HelloReply> responseStream, ServerCallContext context)
+    {
+        // Forward the call on to the greeter service
+        using (var call = _client.SayHellos(request))
```
Comment (scottaddie) [inline aspnetcore/grpc/clientfactory.md]: Same question about this method name.

File: aspnetcore/grpc/clientfactory.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                await responseStream.WriteAsync(call.ResponseStream.Current);
+            }
+        }
+    }
+}
+```
+
+## Configure HttpClient
+
+HttpClientFactory creates the `HttpClient` used by the gRPC client. Standard HttpClientFactory methods can be used to add outgoing request middleware or configure the underlying `HttpClientHandler` of the `HttpClient`:
```
Comment (scottaddie) [inline aspnetcore/grpc/clientfactory.md]:
```suggestion
`HttpClientFactory` creates the `HttpClient` used by the gRPC client. Standard `HttpClientFactory` methods can be used to add outgoing request middleware or to configure the underlying `HttpClientHandler` of the `HttpClient`:
```

File: aspnetcore/grpc/clientfactory.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    })
+    .ConfigurePrimaryHttpMessageHandler(() =>
+    {
+        var handler = new HttpClientHandler();
+        handler.ClientCertificates.Add(LoadCertificate());
+        return handler;
+    });
+```
+
+For more information, see [Make HTTP requests using IHttpClientFactory](fundamentals/http-requests).
```
Comment (scottaddie) [inline aspnetcore/grpc/clientfactory.md]:
```suggestion
For more information, see [Make HTTP requests using IHttpClientFactory](xref:fundamentals/http-requests).
```

File: aspnetcore/grpc/clientfactory.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+## Configure Channel and Interceptors
+
+gRPC specific methods are available to configure gRPC client's underlying channel, or add `Interceptor` instances that the client should use when making gRPC calls.
```
Comment (scottaddie) [inline aspnetcore/grpc/clientfactory.md]:
```suggestion
gRPC-specific methods are available to:

* Configure a gRPC client's underlying channel.
* Add `Interceptor` instances that the client should use when making gRPC calls.
```

File: aspnetcore/grpc/clientfactory.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        o.Credentials = new CustomCredentials();
+    });
+```
+
+## Deadline and cancellation propagation
+
+gRPC clients created by the factory in a gRPC service can be configured with `EnableCallContextPropagation()` to automatically propagate the deadline and cancellation token to child calls.
+
+Call context propagation works by reading those details from the current gRPC request context and automatically propagating them to outgoing calls made by the gRPC client. Call context propagation is an excellent way of ensuring complex, nested gRPC scenarios always propagate the deadline and cancellation.
```
Comment (scottaddie) [inline aspnetcore/grpc/clientfactory.md]:
```suggestion
Call context propagation works by reading those details from the current gRPC request context and automatically propagating them to outgoing calls made by the gRPC client. Call context propagation is an excellent way of ensuring that complex, nested gRPC scenarios always propagate the deadline and cancellation.
```

File: aspnetcore/grpc/configuration.md [hunk 7]
```diff
@@ -33,13 +33,22 @@ Options for a single service override the global options provided in `AddGrpc` a
...
+| `ReceiveMaxMessageSize` | 4 MB | The maximum message size in bytes that can be received by the client. If the server receives a message that exceeds this limit, it throws an exception. Increasing this value allows the server to receive larger messages, but can negatively impact memory consumption. |
+| `TransportOptions` | `null` | Transport options configure how the channel calls the gRPC service. Currently the only implementation is `HttpClientTransport` options lets you specify the `HttpClient` used by gRPC. |
+| `Credentials` | `null` | A `ChannelCredentials` instance. Credentials are used to add authentication metadata to gRPC calls. |
+| `CompressionProviders` | gzip, deflate | A collection of compression providers used to compress and decompress messages. Custom compression providers can be created and added to the collection. The default configured providers support **gzip** and **deflate** compression. |
+
+The following code sets the maximum send and receive message size on the channel, and creates a client:
```
Comment (scottaddie) [inline aspnetcore/grpc/configuration.md]:
```suggestion
The following code:

* Sets the maximum send and receive message size on the channel.
* Creates a client.
```

File: aspnetcore/toc.yml [hunk 8]
```diff
@@ -537,13 +537,17 @@
       uid: grpc/basics
     - name: gRPC services with ASP.NET Core
       uid: grpc/aspnetcore
+    - name: Calling gRPC services with .NET client
```
Comment (scottaddie) [inline aspnetcore/toc.yml]:
```suggestion
    - name: Call gRPC services with .NET client
```

File: aspnetcore/toc.yml [hunk 9]
```diff
@@ -537,13 +537,17 @@
...
+    - name: gRPC client factory integration
+      uid: grpc/clientfactory
     - name: Configuration
       uid: grpc/configuration
     - name: Authentication and authorization
       uid: grpc/authn-and-authz
     - name: Security considerations
       uid: grpc/security
-    - name: Migrating gRPC services from C Core to ASP.NET Core
+    - name: Migrating gRPC services from C Core
```
Comment (scottaddie) [inline aspnetcore/toc.yml]:
```suggestion
    - name: Migrate gRPC services from C Core
```

File: aspnetcore/grpc/client.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
+---
+title: Calling gRPC services with .NET client
+author: jamesnk
+description: Learn how to call gRPC services with the .NET gRPC client.
+monikerRange: '>= aspnetcore-3.0'
+ms.author: jamesnk
+ms.date: 08/06/2019
+uid: grpc/client
+---
+# Calling gRPC services with .NET client
```
Comment (scottaddie) [inline aspnetcore/grpc/client.md]:
```suggestion
# Call gRPC services with the .NET client
```

File: aspnetcore/grpc/client.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
+---
+title: Calling gRPC services with .NET client
```
Comment (scottaddie) [inline aspnetcore/grpc/client.md]:
```suggestion
title: Call gRPC services with the .NET client
```

File: aspnetcore/grpc/client.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+ms.author: jamesnk
+ms.date: 08/06/2019
+uid: grpc/client
+---
+# Calling gRPC services with .NET client
+
+A .NET gRPC client library is available in the [Grpc.Net.Client](https://www.nuget.org/packages/Grpc.Net.Client) NuGet package. This document explains how to:
+
+* Configure gRPC client to call gRPC services
+* Make gRPC calls to unary, server streaming, client streaming and bi-direction streaming methods
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply aspnetcore/grpc/client.md]: Edited your change.

"Server streaming" and "Client streaming" are not hyphenated in any gRPC docs out on the internet. Should be consistent with them.

File: aspnetcore/grpc/clientfactory.md (hunk 1 above)
Comment (JamesNK) [reply aspnetcore/grpc/clientfactory.md]: No, it is streaming multiple hellos.

File: aspnetcore/grpc/client.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+A gRPC client is created from a channel. Start by using `GrpcChannel.ForAddress` to create a channel, and then use the channel to create a gRPC client:
+
+```csharp
+var channel = GrpcChannel.ForAddress("https://localhost:5001");
+var client = new Greet.GreeterClient(channel);
+```
+
+A channel represents a long-lived connection to a gRPC service. When a channel is created it is configured with options related to calling a service. For example, the `HttpClient` used to make calls, the maximum send and receive message size, and logging are options on `GrpcChannelOptions` that can be passed to `GrpcChannel.ForAddress`. For a complete list of options, see [client configuration options](xref:grpc/configuration#configure-client-options).
+
+Creating a channel can be an expensive operation and reusing a channel for gRPC calls offers performance benefits. Multiple gRPC clients can be created from a channel, including different types of clients. Concrete gRPC client types are lightweight objects and can be created when needed.
```
Comment (shirhatti) [inline aspnetcore/grpc/client.md]: Should we add a note here that you can bump against the maximum concurrent streams per connection limit based on how you've configured your server?

File: aspnetcore/grpc/client.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+gRPC clients are concrete client types that are [generated from *\*.proto* files](xref:grpc/basics#generated-c-assets). The concrete gRPC client has methods that translate to the gRPC service in the *\*.proto* file.
+
+A gRPC client is created from a channel. Start by using `GrpcChannel.ForAddress` to create a channel, and then use the channel to create a gRPC client:
+
+```csharp
+var channel = GrpcChannel.ForAddress("https://localhost:5001");
+var client = new Greet.GreeterClient(channel);
+```
+
+A channel represents a long-lived connection to a gRPC service. When a channel is created it is configured with options related to calling a service. For example, the `HttpClient` used to make calls, the maximum send and receive message size, and logging can be specified on `GrpcChannelOptions` and used with `GrpcChannel.ForAddress`. For a complete list of options, see [client configuration options](xref:grpc/configuration#configure-client-options).
```
Comment (shirhatti) [inline aspnetcore/grpc/client.md]: What does client factory do here?

Comment (shirhatti) [reply aspnetcore/grpc/client.md]: Never mind. I see it was addressed in a separate doc

File: aspnetcore/grpc/clientfactory.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+author: jamesnk
+description: Learn how to create gRPC clients using the client factory.
+monikerRange: '>= aspnetcore-3.0'
+ms.author: jamesnk
+ms.date: 08/06/2019
+uid: grpc/clientfactory
+---
+# gRPC client factory integration in .NET Core
+
+gRPC integration with `HttpClientFactory` offers a centralized way to create gRPC clients. It can be used as an alternative to [configuring stand-alone gRPC client instances](xref:grpc/client). Factory integration is available in the [Grpc.Net.ClientFactory](https://www.nuget.org/packages/Grpc.Net.ClientFactory) NuGet package.
```
Comment (shirhatti) [inline aspnetcore/grpc/clientfactory.md]: Mention both variants of the factory?

File: aspnetcore/tutorials/grpc/grpc-start/sample/GrpcGreeterClient/GrpcGreeterClient.csproj [hunk 16]
```diff
@@ -7,8 +7,8 @@
 
   <ItemGroup>
     <PackageReference Include="Google.Protobuf" Version="3.8.0" />
-    <PackageReference Include="Grpc.Net.Client" Version="0.1.22-pre2" />
-    <PackageReference Include="Grpc.Tools" Version="1.22.0-pre1" PrivateAssets="All" />
+    <PackageReference Include="Grpc.Net.Client" Version="0.2.23-dev201908080701" />
```
Comment (shirhatti) [inline aspnetcore/tutorials/grpc/grpc-start/sample/GrpcGreeterClient/GrpcGreeterClient.csproj]: I'm assuming we don't plan to publish this till preview9 ships?

Comment (JamesNK) [reply aspnetcore/tutorials/grpc/grpc-start/sample/GrpcGreeterClient/GrpcGreeterClient.csproj]: ye

File: aspnetcore/grpc/client.md (hunk 13 above)
Comment (JamesNK) [reply aspnetcore/grpc/client.md]: I want to investigate that more before writing guidance.

**Pull request merged**